contracts/LinniaRecords.sol

```diff
@@ -162,40 +174,43 @@ contract LinniaRecords is Ownable {
     /* Internal functions */
 
     function _addRecord(
-        bytes32 fileHash, address fileOwner, string keywords, bytes32 ipfsHash)
+        bytes32 dataHash, address owner, string metadata, bytes32 dataUri)
         internal
-        returns (bool)
+        returns (bytes32)
     {
         // validate input
-        require(fileHash != 0 && ipfsHash != 0);
+        require(dataHash != 0 && dataUri != 0);
+        // calculate root hash
+        bytes32 metaHash = keccak256(metadata);
+        bytes32 rootHash = keccak256(dataHash, metaHash);
```

Comment:
`bytes32 rootHash = keccak256(dataHash, metaHash);`

should be refactored into `bytes32 _rootHash = rootHash(dataHash, metaHash);` and a set of functions on the main contract.

```
    function rootHash(Record record) 
        private
        pure
        returns (bytes32)
    {
        return rootHash(record.dataHash, record.metaHash);
    }

    function rootHash(bytes32 dataHash, bytes32 metaHash) 
        public
        pure
        returns (bytes32)
    {
        return keccak256(abi.encodePacked(dataHash, metaHash));
    }
```

This is an important piece of standalone functionality and if it changes, it should be changed in a single place instead of all over.

I would also seriously consider making `Record` a standalone contract. It has lots of info and it's easier to pass around. It would then be expected for someone to be able to do: `record.rootHash()`.

Overall there's a big problem here. The access pattern with the previous version was `Record r = hub.recordsContract.records(fileHash)` which required me to know a single piece of information - `fileHash`. Now I need to know two pieces of information and the code above becomes:  `Record r = hub.recordsContract.records(keccak256(abi.encodePacked(dataHash, metaHash)))`. This isn't a step forward in terms of usability.

Reply:
I think this allows users to upload the same piece of data with different metadata, and having them attested differently. Maybe add a mapping from data hash to a list of root hashes?